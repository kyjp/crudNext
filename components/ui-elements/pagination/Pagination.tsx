import React, { FC, MouseEvent } from 'react'

type PaginationType = {
    currentPage: number
    limit: number
    count: number
    onClick: (event: MouseEvent<HTMLButtonElement>) => void
}

const Pagination: FC<PaginationType> = ({
    currentPage,
    limit,
    count,
    onClick
}) => {
    const totalPages = Math.ceil(count / limit)
    let startPage = Math.max(1, currentPage - 2)
    let endPage = Math.min(totalPages, currentPage + 2)

    if(currentPage <= 3) {
        endPage = Math.min(5, totalPages)
    } else if(totalPages - 2 <= currentPage) {
        startPage = totalPages - 4
    }

    const pageNumbers = []
    for(let i = startPage; i <= endPage; i++) {
        pageNumbers.push(i)
    }

    return (
        <div className="flex justify-center items-center -space-x-px h-8 text-sm">
            <button
                className="flex items-center justify-center px-3 h-8 ms-0 leading-tight text-gray-500 bg-white border border-e-0 border-gray-300 rounded-s-lg hover:bg-gray-100 hover:text-gray-700"
                onClick={onClick}
            >
                <span className="sr-only">Previous</span>
                <svg className="w-2.5 h-2.5 rtl:rotate-180" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 6 10">
                    <path stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="M5 1 1 5l4 4"/>
                </svg>
            </button>
            {pageNumbers.map((item, index) => 
                <button
                    className={`flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700${currentPage == item ? ' font-bold' : ''}`}
                    key={`pagination-${index}`}
                    onClick={onClick}
                >
                    {item}
                </button>
            )}
            <button
                className="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 rounded-e-lg hover:bg-gray-100 hover:text-gray-700"
                onClick={onClick}
            >
                <span className="sr-only">Next</span>
                <svg className="w-2.5 h-2.5 rtl:rotate-180" aria-hidden="true" xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 6 10">
                    <path stroke="currentColor" strokeLinecap="round" strokeLinejoin="round" strokeWidth="2" d="m1 9 4-4-4-4"/>
                </svg>
            </button>
        </div>
    )
}

export default Pagination