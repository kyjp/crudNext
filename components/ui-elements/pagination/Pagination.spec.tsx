import { fireEvent, render, screen } from "@testing-library/react"
import Pagination from "./Pagination"
import '@testing-library/jest-dom'

describe('Test group of Pagenation component', () => {
    const onClickMock = vi.fn()
    afterEach(() => {
        onClickMock.mockClear()
    })
    test('Snapshot Test > renders Pagination component', () => {
        const {asFragment} = render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        expect(asFragment()).toMatchSnapshot()
    })

    test("should highlight the current page", async () => {
        render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        const currentPageButton = screen.getByText('3')
        expect(currentPageButton).toHaveClass('font-bold')
    })

    test('should call onClick when a page number is clicked', () => {
        render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        const pageButton = screen.getByText('3')
        fireEvent.click(pageButton)
        expect(onClickMock).toHaveBeenCalledTimes(1)
    })

    test('should call onClick when the next button is clicked', () => {
        render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        const prevButton = screen.getByText('Previous').closest('button') as HTMLButtonElement
        fireEvent.click(prevButton)
        expect(onClickMock).toHaveBeenCalledTimes(1)
    })

    test('should call onClick when the next button is called', () => {
        render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        const nextButton = screen.getByText('Next').closest('button') as HTMLButtonElement
        fireEvent.click(nextButton)
        expect(onClickMock).toHaveBeenCalledTimes(1)
    })

    test('shold render previous and next buttons', () => {
        render(
            <Pagination
                currentPage={3}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        expect(screen.getByText('Previous').closest('button') as HTMLButtonElement).toBeInTheDocument()
        expect(screen.getByText('Next').closest('button') as HTMLButtonElement).toBeInTheDocument()
    })

    test('should adjust page numbers when at the beginning of the list', () => {
        render(
            <Pagination
                currentPage={1}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        expect(screen.getByText('1')).toBeInTheDocument()
        expect(screen.getByText('2')).toBeInTheDocument()
        expect(screen.getByText('3')).toBeInTheDocument()
        expect(screen.getByText('4')).toBeInTheDocument()
        expect(screen.getByText('5')).toBeInTheDocument()
    })

    test('should adjust page numbers when at the end of the list', () => {
        render(
            <Pagination
                currentPage={5}
                limit={10}
                count={50}
                onClick={onClickMock}
            />
        )
        expect(screen.getByText('3')).toBeInTheDocument()
        expect(screen.getByText('4')).toBeInTheDocument()
        expect(screen.getByText('5')).toBeInTheDocument()
    })
})