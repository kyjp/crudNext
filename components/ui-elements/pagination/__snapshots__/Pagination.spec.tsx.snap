// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Snapshot Test > renders Pagination component 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center -space-x-px h-8 text-sm"
  >
    <button
      class="flex items-center justify-center px-3 h-8 ms-0 leading-tight text-gray-500 bg-white border border-e-0 border-gray-300 rounded-s-lg hover:bg-gray-100 hover:text-gray-700"
    >
      <span
        class="sr-only"
      >
        Previous
      </span>
      <svg
        aria-hidden="true"
        class="w-2.5 h-2.5 rtl:rotate-180"
        fill="none"
        viewBox="0 0 6 10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 1 1 5l4 4"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      1
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      2
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700 font-bold"
    >
      3
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      4
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      5
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 rounded-e-lg hover:bg-gray-100 hover:text-gray-700"
    >
      <span
        class="sr-only"
      >
        Next
      </span>
      <svg
        aria-hidden="true"
        class="w-2.5 h-2.5 rtl:rotate-180"
        fill="none"
        viewBox="0 0 6 10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m1 9 4-4-4-4"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Test group of Pagenation component > Snapshot Test > renders Pagination component 1`] = `
<DocumentFragment>
  <div
    class="flex justify-center items-center -space-x-px h-8 text-sm"
  >
    <button
      class="flex items-center justify-center px-3 h-8 ms-0 leading-tight text-gray-500 bg-white border border-e-0 border-gray-300 rounded-s-lg hover:bg-gray-100 hover:text-gray-700"
    >
      <span
        class="sr-only"
      >
        Previous
      </span>
      <svg
        aria-hidden="true"
        class="w-2.5 h-2.5 rtl:rotate-180"
        fill="none"
        viewBox="0 0 6 10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 1 1 5l4 4"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      1
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      2
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700 font-bold"
    >
      3
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      4
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 hover:bg-gray-100 hover:text-gray-700"
    >
      5
    </button>
    <button
      class="flex items-center justify-center px-3 h-8 leading-tight text-gray-500 bg-white border border-gray-300 rounded-e-lg hover:bg-gray-100 hover:text-gray-700"
    >
      <span
        class="sr-only"
      >
        Next
      </span>
      <svg
        aria-hidden="true"
        class="w-2.5 h-2.5 rtl:rotate-180"
        fill="none"
        viewBox="0 0 6 10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m1 9 4-4-4-4"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
